astropy/table/table.py

```diff
@@ -1250,7 +1257,12 @@ def __setitem__(self, item, value):
 
             if isinstance(item, six.string_types):
                 # Set an existing column
-                self.columns[item][:] = value
+                try:
+                    # See definition of mask property for discussion of next line.
+                    assert not getattr(self, '_setitem_inplace', False)
```

Comment:
An `assert` is fine if it tests something that is expected to be true -- it would seem OK to omit that test when one optimizes -- but not when it is used to decide which logical path to take. So, I guess OK except in try/except clauses.

That said, obviously it is better to do
```
if <something-unlikely>:
    raise ShouldNotHappenError('...')
```